### PR TITLE
libbpf-rs: add sk_lookup

### PR DESCRIPTION
**Pull request by alpire**
This PR adds a `attach_sk_lookup` in order to implement [sk_loopup programs](https://www.kernel.org/doc/html/latest/bpf/prog_sk_lookup.html).

### CODE REVIEW

**Review by mimullin-bbry — APPROVED** [review]
+1

File: libbpf-rs/src/program.rs [hunk 1]
```diff
@@ -344,4 +354,23 @@ impl Program {
             Ok(Link::new(ptr))
         }
     }
+
+    /// Attach this program to [sk_lookup](https://lwn.net/Articles/819618/)
+    pub fn attach_sk_lookup(&mut self, target_fd: i32) -> Result<()> {
+        let mut opts = libbpf_sys::bpf_link_create_opts::default();
+        opts.sz = std::mem::size_of::<libbpf_sys::bpf_link_create_opts>() as u64;
+        let fd = unsafe {
```
Comment (insearchoflosttime) [inline libbpf-rs/src/program.rs:362]: Do we need to return back the fd to the user so that the bpf link may be closed when the user is done with it?

Comment (anakryiko) [reply libbpf-rs/src/program.rs:362]: Yeah, It has to return Result<Link>

But I also see that libbpf provides bpf_program__attach_netns() API for this, so let's stay consistent?

Comment (alpire) [reply libbpf-rs/src/program.rs]: This is much better indeed. I just pushed an update to rely on `bpf_program__attach_netns` & return a `Result<Link>`. Thanks for the help!

File: libbpf-rs/src/program.rs [hunk 2]
```diff
@@ -110,6 +111,15 @@ pub enum ProgramAttachType {
...
     LsmMac,
+    TraceIter,
+    CgroupInet4Getpeername,
+    CgroupInet6Getpeername,
+    CgroupInet4Getsockname,
+    CgroupInet6Getsockname,
+    XdpDevmap,
+    CgroupInetSockRelease,
+    XdpCpumap,
+    SkLookup,
```
Comment (anakryiko) [inline libbpf-rs/src/program.rs:122]: while you are at it, why not add the rest of values:

```
 994 ▸       BPF_SK_SKB_VERDICT,¬
 995 ▸       BPF_SK_REUSEPORT_SELECT,¬
 996 ▸       BPF_SK_REUSEPORT_SELECT_OR_MIGRATE,¬
 997 ▸       BPF_PERF_EVENT,¬
```

**Review by anakryiko — APPROVED** [review]
LGTM, but if you'd be so nice to complete the bpf_attach_type enum, that would be great. Thanks!

File: libbpf-rs/src/program.rs [hunk 3]
```diff
@@ -73,6 +73,7 @@ pub enum ProgramType {
     StructOps,
     Ext,
     Lsm,
+    SkLookup,
```
Comment (anakryiko) [inline libbpf-rs/src/program.rs:76]: same here, we now also have BPF_PROG_TYPE_SYSCALL, would be good to add

**Comment by anakryiko** [comment]
@alpire please see https://github.com/libbpf/libbpf-rs/pull/141#pullrequestreview-803252416 as well. A bit cleaner and fuller commit message wouldn't hurt. Signed-off-by is also a good style borrowed from kernel development process that libbpf uses, so would be good to keep that for libbpf-rs as well. Thanks!

File: libbpf-rs/src/program.rs (hunk 2 above)
Comment (alpire) [reply libbpf-rs/src/program.rs:123]: Sounds good. Done!

File: libbpf-rs/src/program.rs (hunk 3 above)
Comment (alpire) [reply libbpf-rs/src/program.rs:76]: Sounds good. Done!

**Comment by alpire** [comment]
> @alpire please see [#141 (review)](https://github.com/libbpf/libbpf-rs/pull/141#pullrequestreview-803252416) as well. A bit cleaner and fuller commit message wouldn't hurt. Signed-off-by is also a good style borrowed from kernel development process that libbpf uses, so would be good to keep that for libbpf-rs as well. Thanks!

I just pushed a new commit with a more descriptive error message and Signed-off-by. Thanks for the review!

Adding [contributing guidelines](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors#about-contributing-guidelines) might be helpful for newcomers like me so that we can follow the preferred style of the repo.

File: libbpf-rs/src/program.rs [hunk 4]
```diff
@@ -344,4 +360,15 @@ impl Program {
             Ok(Link::new(ptr))
         }
     }
+
+    /// Attach this program to [sk_lookup](https://lwn.net/Articles/819618/)
+    pub fn attach_sk_lookup(&mut self, netns_fd: i32) -> Result<Link> {
```
Comment (anakryiko) [inline libbpf-rs/src/program.rs:365]: wait, I thought you are going to add "attach_netns" method... Why deviate from the libbpf naming? sk_lookup is just one subtype of programs that can be attached here. I think flow_dissector is the other one right now? This method should support both out of the box, no? So sk_lookup-specific name is suboptimal.

**Review by anakryiko — CHANGES_REQUESTED** [review]

**Comment by anakryiko** [comment]
> > @alpire please see [#141 (review)](https://github.com/libbpf/libbpf-rs/pull/141#pullrequestreview-803252416) as well. A bit cleaner and fuller commit message wouldn't hurt. Signed-off-by is also a good style borrowed from kernel development process that libbpf uses, so would be good to keep that for libbpf-rs as well. Thanks!
> 
> I just pushed a new commit with a more descriptive error message and Signed-off-by. Thanks for the review!

Yeah, looks good (except I just noticed that you didn't use "attach_netns" terminology.

> 
> Adding [contributing guidelines](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors#about-contributing-guidelines) might be helpful for newcomers like me so that we can follow the preferred style of the repo.

Yeah, good idea. I'll create an issue for this. It would be very meta to have a contributor contribute contributing guidelines ;)

File: libbpf-rs/src/program.rs (hunk 4 above)
Comment (alpire) [reply libbpf-rs/src/program.rs:365]: Oops, sorry, I misunderstood your previous comment. The method is now called `attach_netns`. I agree this is better.

Comment (alpire) [reply libbpf-rs/src/program.rs:365]: Oh, the MR is already merged. I'll open a new one to rename the method.

Comment (anakryiko) [reply libbpf-rs/src/program.rs:365]: I'll fix it up locally, don't send another one

**Pull request merged**